src/libpmempool/replica.c

```diff
@@ -528,29 +658,469 @@ check_checksums_and_signatures(struct pool_set *set,
+ *                                         recovery file before clearing them
+ */
+static int
+replica_badblocks_recovery_file_save(struct part_health_status *part_hs)
+{
+	LOG(3, "part_health_status %p", part_hs);
+
+	ASSERTeq(part_hs->recovery_file_exists, 1);
+	ASSERTne(part_hs->recovery_file_name, NULL);
+
+	struct badblocks *bbs = &part_hs->bbs;
+	char *path = part_hs->recovery_file_name;
+	int ret = -1;
+
+	int fd = os_open(path, O_WRONLY | O_TRUNC);
+	if (fd < 0) {
+		ERR("!opening bad block recovery file failed -- '%s'", path);
+		return -1;
+	}
+
+	FILE *recovery_file_name = os_fdopen(fd, "w");
+	if (recovery_file_name == NULL) {
+		ERR(
+			"!opening a file stream for bad block recovery file failed -- '%s'",
+			path);
+		os_close(fd);
+		return -1;
+	}
+
+	/* save bad blocks */
+	for (unsigned i = 0; i < bbs->bb_cnt; i++) {
+		fprintf(recovery_file_name, "%llu %u\n",
+			bbs->bbv[i].offset, bbs->bbv[i].length);
+	}
+
+	if (fflush(recovery_file_name) == EOF) {
+		ERR("!flushing bad block recovery file failed -- '%s'", path);
+		goto exit_error;
+	}
+
+	if (os_fsync(fd) < 0) {
+		ERR("!syncing bad block recovery file failed -- '%s'", path);
+		goto exit_error;
+	}
+
+	/* save the finish flag */
+	fprintf(recovery_file_name, "0 0\n");
+
+	if (fflush(recovery_file_name) == EOF) {
+		ERR("!flushing bad block recovery file failed -- '%s'", path);
+		goto exit_error;
+	}
+
+	if (os_fsync(fd) < 0) {
+		ERR("!syncing bad block recovery file failed -- '%s'", path);
+		goto exit_error;
+	}
+
+	LOG(3, "bad blocks saved in the recovery file -- '%s'", path);
+	ret = 0;
+
+exit_error:
+	os_fclose(recovery_file_name);
+
+	return ret;
+}
+
+/*
+ * replica_part_badblocks_recovery_file_read -- read bad blocks
+ *                                             from the bad block recovery file
+ *                                             for the current part
+ */
+static int
+replica_part_badblocks_recovery_file_read(struct part_health_status *part_hs)
+{
+	LOG(3, "part_health_status %p", part_hs);
+
+	ASSERT(part_hs->recovery_file_exists);
+	ASSERTne(part_hs->recovery_file_name, NULL);
+
+	VEC(bbsvec, struct bad_block) bbv = VEC_INITIALIZER;
+	char *path = part_hs->recovery_file_name;
+	struct bad_block bb;
+	int ret = -1;
+
+	FILE *recovery_file = os_fopen(path, "r");
+	if (!recovery_file) {
+		ERR("!opening the recovery file for reading failed -- '%s'",
+			path);
+		return -1;
+	}
+
+	unsigned long long min_offset = 0; /* minimum possible offset */
+
+	do {
+		if (fscanf(recovery_file, "%llu %u\n",
+				&bb.offset, &bb.length) < 2) {
+			LOG(1, "incomplete bad block recovery file -- '%s'",
+				path);
+			ret = 1;
+			goto error_exit;
+		}
+
+		if (bb.offset == 0 && bb.length == 0) {
+			/* finish_flag */
+			break;
+		}
+
+		/* check if bad blocks build an increasing sequence */
+		if (bb.offset < min_offset) {
+			ERR(
+				"wrong format of bad block recovery file (bad blocks do not build an increasing sequence) -- '%s'",
```

Comment:
(bad blocks are not sorted by the offset in ascending order)

Reply:
done